### PR TITLE
Add support for mining ommers, up to 8 blocks behind the head job

### LINKED ISSUE #2572: block creator: allow submitting 8-block deep uncle solutions
### Description
Besu is capable of mining the latest block only. Like in Geth, it should also accept 8-block deep uncles besides the latest pending block.

```java
public boolean submitSolution(final PoWSolution solution) {
    final Optional<PoWSolverJob> jobSnapshot = currentJob;
    if (jobSnapshot.isEmpty()) {
      LOG.debug("No current job, rejecting miner work");
      return false;
    }

    final PoWSolverJob job = jobSnapshot.get();
    final PoWSolverInputs inputs = job.getInputs();
    if (!inputs.getPrePowHash().equals(solution.getPowHash())) {
      // ! REJECTS POTENTIAL UNCLE HERE
      LOG.debug("Miner's solution does not match current job");
      return false;
    }
    final Optional<PoWSolution> calculatedSolution = testNonce(inputs, solution.getNonce());

    if (calculatedSolution.isPresent()) {
      LOG.debug("Accepting a solution from a miner");
      currentJob.get().solvedWith(calculatedSolution.get());
      return true;
    }
    LOG.debug("Rejecting a solution from a miner");
    return false;
  }
```

### PR DESCRIPTION
**Pull request by atoulme**
Signed-off-by: Antoine Toulme <antoine@lunar-ocean.com>

## PR description
This allows miners to send solutions for blocks up to 8 blocks behind head, to be included as ommers.

## Fixed Issue(s)
Fixes #2572

## Changelog

- [ ] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/PoWSolver.java [hunk 1]
```diff
@@ -26,10 +26,14 @@
 
 import com.google.common.base.Stopwatch;
 import org.apache.logging.log4j.Logger;
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.concurrent.ExpiringMap;
 import org.apache.tuweni.units.bigints.UInt256;
 
 public class PoWSolver {
 
+  private static final long POW_JOB_TTL = 1000 * 60 * 5; // 5 minutes
```
Comment (matkt) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/PoWSolver.java:35]: is this value related to something ?

Comment (atoulme) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/PoWSolver.java:35]: Not really. 13s per block on average, so 13 times 8 is 108s, times 2 makes 3 minutes and a half, and buffered to 5 minutes to ensure we catch stragglers. We can use a different approach. I have another ticket to add meters to count accepted solutions vs rejected solutions.

Comment (matkt) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/PoWSolver.java:35]: ok it was to make sure I didn't miss anything. did you had the opportunity to test it ? I can test it on ropsten if you want

Comment (atoulme) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/PoWSolver.java:35]: I did not. I kinda wanted folks to review as well. I added some unit tests to make sure it made sense.

Comment (shemnon) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/PoWSolver.java:35]: My instinct says this should be configurable.  Probably via an `--X` flag.

Comment (atoulme) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/PoWSolver.java:35]: Is 5 minutes an ok default?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/PoWSolver.java [hunk 2]
```diff
@@ -149,22 +155,50 @@ public void cancel() {
...
+      return false;
+    }
+    final PoWSolverInputs inputs = jobToTestWith.getInputs();
+
     final Optional<PoWSolution> calculatedSolution = testNonce(inputs, solution.getNonce());
 
     if (calculatedSolution.isPresent()) {
       LOG.debug("Accepting a solution from a miner");
-      currentJob.get().solvedWith(calculatedSolution.get());
+      currentJobs.remove(solution.getPowHash());
```
Comment (shemnon) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/PoWSolver.java:200]: Does the miner want multiple ommers minable?  Removing on the first solution precludes that.

Comment (atoulme) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/PoWSolver.java:200]: Right now, we only accept one solution per block (the `solveFor` method blocks until the future completes). This repeats the same mechanism but with ommer blocks. We could potentially choose to allow more than one solution per block, turn those into ommers and replace the proposed block with the block with the best nonce. That's why I wanted a review, I'm just not sure how far to push this.

Comment (atoulme) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/PoWSolver.java:200]: To be clear, that'd be best addressed as a separate PR, because changing `solveFor` feels like a change worth doing on its own, whether we accept mining ommers or not.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/PoWSolver.java (hunk 1 above)
Comment (shemnon) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/PoWSolver.java:35]: maybe 2x high?  But since we are just keeping block headers (right?) from 5 min ago unless things get _very_ hairy I don't see it causing resource problems.  If we were keeping TXes it could be problematic.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/PoWSolver.java [hunk 3]
```diff
@@ -149,22 +155,50 @@ public void cancel() {
...
     if (jobSnapshot.isEmpty()) {
       LOG.debug("No current job, rejecting miner work");
       return false;
     }
 
-    final PoWSolverJob job = jobSnapshot.get();
-    final PoWSolverInputs inputs = job.getInputs();
-    if (!inputs.getPrePowHash().equals(solution.getPowHash())) {
-      LOG.debug("Miner's solution does not match current job");
+    PoWSolverJob headJob = jobSnapshot.get();
```
Comment (garyschulte) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/PoWSolver.java:164]: might as well get rid of jobSnapshot reference.

**Comment by shemnon** [comment]
Plumb the `--X` flag through to configure retention and I'll approve it unless someone from ConsenSys objects.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/PoWSolver.java (hunk 1 above)
Comment (atoulme) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/PoWSolver.java:35]: Yep, only block headers. It should be fine.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/PoWSolver.java [hunk 4]
```diff
@@ -26,11 +26,15 @@
 
 import com.google.common.base.Stopwatch;
 import org.apache.logging.log4j.Logger;
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.concurrent.ExpiringMap;
 import org.apache.tuweni.units.bigints.UInt256;
 
 public class PoWSolver {
 
+  private static final int MAX_OMMER_DEPTH = 8;
```
Comment (shemnon) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/PoWSolver.java:35]: For another PR, another `--X` candidate.

**Review by shemnon — APPROVED** [review]

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/PoWSolver.java (hunk 4 above)
Comment (atoulme) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/PoWSolver.java:35]: https://github.com/hyperledger/besu/pull/2584 is open
